s3/src/main/scala/akka/stream/alpakka/s3/auth/CanonicalRequest.scala

```diff
@@ -43,14 +65,21 @@ private[alpakka] object CanonicalRequest {
    * percent encoded path delimiters back to their decoded counterparts.
    */
   private def preprocessPath(path: Path): String =
-    uriEncode(path.toString).replace(":", "%3A").replace("%2F", "/")
+    uriEncode(removeRedundantAndRelativePath(path.toString))
+      .replace(":", "%3A")
+      .replace("%2F", "/")
+      .replace("%7E", "~")
+      .replace("+", "%20")
 
   def canonicalHeaderString(headers: Seq[HttpHeader]): String = {
     val grouped = headers.groupBy(_.lowercaseName())
```

Comment:
This seems to be missing the value of `req.entity.contentType` which is encoded specially in the akka-http model.